javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,225 @@
+import javascript
+
+module EventEmitter {
+  /** Gets the name of a method on `EventEmitter` that returns `this`. */
+  string chainableMethod() {
+    result = "off" or
+    result = "removeAllListeners" or
+    result = "removeListener" or
+    result = "setMaxListeners" or
+    result = on()
+  }
+
+  /** Gets the name of a method on `EventEmitter` that registers an event handler. */
+  string on() {
+    result = "addListener" or
+    result = "on" or
+    result = "once" or
+    result = "prependListener" or
+    result = "prependOnceListener"
+  }
+
+  /**
+   * Gets a node that refers to an EventEmitter object.
+   */
+  DataFlow::SourceNode trackEventEmitter(EventEmitter::Range emitter) {
+    result = trackEventEmitter(DataFlow::TypeTracker::end(), emitter)
+  }
+
+  private DataFlow::SourceNode trackEventEmitter(
+    DataFlow::TypeTracker t, EventEmitter::Range emitter
+  ) {
+    t.start() and result = emitter
+    or
+    exists(DataFlow::TypeTracker t2, DataFlow::SourceNode pred |
+      pred = trackEventEmitter(t2, emitter)
+    |
+      result = pred.track(t2, t)
+      or
+      // invocation of a chainable method
+      exists(DataFlow::MethodCallNode mcn |
+        mcn = pred.getAMethodCall(EventEmitter::chainableMethod()) and
+        // exclude getter versions
+        exists(mcn.getAnArgument()) and
+        result = mcn and
+        t = t2.continue()
+      )
+    )
+  }
+
+  /**
+   * An object that implements the EventEmitter API.
+   * Extending this class does nothing, its mostly to indicate intent.
+   * The magic only happens when extending EventRegistration::Range and EventDispatch::Range.
```

Comment:
>    * The magic only happens when extending EventRegistration::Range and EventDispatch::Range.

Can we avoid the word `magic`?
Is it correct that this does nothing? `trackEventEmitter` performs tracking of all `Range` instances at the moment..

Can we briefly mention in this qldoc how the `EventRegistration` and `EventDispatch` instances relate conceptually? At the very least, the "registration" and "dispatch" concepts should be explained somewhere in this module - I do not think they are at the moment.

Reply:
> Is it correct that this does nothing? `trackEventEmitter` performs tracking of all `Range` instances at the moment..

Correct, it pretty much does nothing.   
I changed `trackEventEmitter` to track `EventEmitter` instead of `EventEmitter::Range`.